boostregex/AnsiDocumentIterator.h

```diff
@@ -1,13 +1,35 @@
-﻿#ifndef ANSIDOCUMENTITERATOR_H_12481491281240
+﻿// This file is part of Notepad++ project
+// Copyright (C) 2021 Notepad++ authors.
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// at your option any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+#ifndef ANSIDOCUMENTITERATOR_H_12481491281240
 #define ANSIDOCUMENTITERATOR_H_12481491281240
 
 #include "Position.h"
 
 namespace Scintilla {
 
-class AnsiDocumentIterator : public std::iterator<std::bidirectional_iterator_tag, char>
+class AnsiDocumentIterator
 {
 public:
+	using iterator_category = std::bidirectional_iterator_tag;
+	using value_type = char;
+	using difference_type = ptrdiff_t;
+	using pointer = char*;
+	using reference = char&;
+

```

Comment:
Any explanation for using `using ...` to replace  ` public std::iterator<std::bidirectional_iterator_tag, char>` please?

Reply:
`std::iterator` was deprecated in C++17.
It is [defined](https://en.cppreference.com/w/cpp/iterator/iterator) as following:

```
template< 
    typename Category,
    typename T,
    typename Distance = std::ptrdiff_t,
    typename Pointer = T*,
    typename Reference = T& 
> struct iterator;
```

It was deprecated mainly because it wasn't clear which type is what when specifying template parameters.
Using declaration allow to be more specific.

> What is deprecated is the technique of inheriting from std::iterator to define those types. That’s it.
> 
```
class MyOutputIterator : public std::iterator<std::output_iterator_tag, void, void, void, void>
{
```
> This code doesn’t say to which type of the interface (value_type, reference_type…) each of the types passed corresponds.
> 
> A more explicit way to go about it is to write the using declarations (or typedefs if you’re before C++11) directly inside of the iterator

More info:
https://www.fluentcpp.com/2018/05/08/std-iterator-deprecated/

Originally, I just wanted to suppress the warning. But @chcg suggested to modify the code so that the warning isn't generated and the code is a valid C++17 code.

Reply:
Thank you for the explanation. However, we cannot just count on this article.
Has this modification been tested? If yes, how?

Reply:
Yes, I did some testing. I opened the file 
[lorem.txt](https://github.com/notepad-plus-plus/notepad-plus-plus/files/6682034/lorem.txt)
I checked auto complete and regex search/count with simple expressions like `q\w+s\b`. 

What else proofs are needed?
https://en.cppreference.com/w/cpp/iterator/iterator
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0174r2.html

The generated assembler is exactly the same.
The only difference is that _with_ `std::iterator` the type alias (using/typedef) is defined in the base class.
And _without_ `std::iterator` the type aliases are defined in the same class (AnsiDocumentIterator).

Below is a short example.
With iterator base:
https://godbolt.org/z/xvffq8EnW
Without iterator base:
https://godbolt.org/z/bMGs9aPqz
If you do a diff of the generated assembly, there is no difference.